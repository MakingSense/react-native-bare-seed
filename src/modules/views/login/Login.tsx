import React from 'react';
import { View, Text, TextInput, KeyboardAvoidingView } from 'react-native';
import { NavigationScreenProp, NavigationScreenOptions } from 'react-navigation';
import { Button } from 'react-native-elements';

import { ENV, STYLE } from '../../../constants';
import styles from './styles';

export interface ILoginProps {
  isLoading: boolean;
  hasError: boolean;
  navigation: NavigationScreenProp<any, any>;
  login: (email: string, password: string) => void;
  setNavigation: (navigation: NavigationScreenProp<any, any>) => void;
  checkForUpdates: () => void;
}

export interface ILoginState {
  email: string;
  password: string;
}

export default class Login extends React.PureComponent<ILoginProps, ILoginState> {
  public static navigationOptions: NavigationScreenOptions = {
    title: 'Login'
  };

  public static isAndroid: boolean = ENV.PLATFORM.IS_ANDROID;

  public state: ILoginState = {
    email: 'dgeslin@makingsense.com',
    password: 'password'
  };

  public componentDidMount() {
    const { setNavigation, navigation, checkForUpdates } = this.props;
    setNavigation(navigation);
    checkForUpdates();
  }

  public render() {
    const { isLoading, hasError } = this.props;
    const { email, password } = this.state;

    return (
      <View style={styles.loginContainer}>
        <Text style={styles.heading1}>Making Sense RN seed</Text>
        <Text style={styles.heading2}>Please login to your account</Text>
        {hasError ? <Text style={styles.headingError}>{'Oops! Invalid email or password'}</Text> : null}
        <KeyboardAvoidingView behavior="padding" style={styles.keyboardContainer} keyboardVerticalOffset={Login.isAndroid ? ENV.KEYBOARD_VERTICAL_OFFSET : 0}>
          <View style={styles.inputTemplate}>
            <View style={styles.inputContainer}>
              <Text style={styles.inputLabel}>{'EMAIL:'}</Text>
              <TextInput
                keyboardType="email-address"
                placeholder="Enter your email..."
                placeholderTextColor={STYLE.COLOR.OCTANARY}
                style={[styles.input, hasError && styles.inputError]}
                underlineColorAndroid="rgba(0, 0, 0, 0)"
                value={email}
                onChangeText={value => this.handleOnChange('email', value)}
              />
            </View>
            <View style={styles.inputContainer}>
              <Text style={styles.inputLabel}>{'Password:'.toUpperCase()}</Text>
              <TextInput
                placeholder="Enter your password..."
                placeholderTextColor={STYLE.COLOR.OCTANARY}
                secureTextEntry={true}
                style={[styles.input, hasError && styles.inputError]}
                underlineColorAndroid="rgba(0, 0, 0, 0)"
                value={password}
                onChangeText={value => this.handleOnChange('password', value)}
              />
            </View>
          </View>
        </KeyboardAvoidingView>
        <Button loading={isLoading} disabled={isLoading} onPress={this.login} title="Login" testID="login-button" />
      </View>
    );
  }

  private handleOnChange(field: 'email' | 'password', value: string) {
    this.setState({ [field]: value } as any);
  }

  private login = () => {
    const { email, password } = this.state;
    const { login } = this.props;
    login(email, password);
  };
}
