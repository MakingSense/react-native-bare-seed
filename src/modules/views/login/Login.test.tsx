import React from 'react';
import Login, { ILoginProps } from './Login';
import { shallow } from 'enzyme';

import renderer from 'react-test-renderer';

describe('Login', () => {
  let props: ILoginProps;
  beforeEach(() => {
    props = {
      isLoading: false,
      hasError: false,
      login: jest.fn(),
      setNavigation: jest.fn(),
      navigation: jest.fn() as any,
      checkForUpdates: jest.fn()
    };
  });

  it('renders without crashing', () => {
    const rendered = renderer.create(<Login {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('renders without crashing on Android', () => {
    Login.isAndroid = true;
    const rendered = renderer.create(<Login {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
    Login.isAndroid = false;
  });

  it('renders showing errors', () => {
    const rendered = renderer.create(<Login {...{ ...props, hasError: true }} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should call initial functions when session is NOT checked mounted', () => {
    const wrapper = shallow(<Login {...props} />);
    wrapper.update();
    expect(props.setNavigation).toBeCalledWith(props.navigation);
    expect(props.checkForUpdates).toBeCalled();
  });

  it('should trigger a login', () => {
    const email = 'email';
    const password = 'password';
    const wrapper = shallow(<Login {...props} />);
    wrapper.setState({ email, password });
    (wrapper.find('[testID="login-button"]').prop('onPress') as any)();
    expect(props.login).toBeCalledWith(email, password);
  });

  it('should handleOnChange', () => {
    const email = 'email';
    const password = 'password';
    const wrapper = shallow(<Login {...props} />);
    (wrapper
      .find('TextInput')
      .first()
      .prop('onChangeText') as any)(email);
    (wrapper
      .find('TextInput')
      .last()
      .prop('onChangeText') as any)(password);
    const prevState = JSON.parse(JSON.stringify(wrapper.instance().state));
    expect(wrapper.instance().state).toEqual({ ...prevState, email, password });
  });
});
