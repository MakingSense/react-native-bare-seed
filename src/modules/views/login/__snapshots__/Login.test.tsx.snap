// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders showing errors 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#1745D5",
        "fontSize": 23,
        "fontWeight": "600",
        "marginBottom": 8,
        "marginTop": 16,
        "textAlign": "left",
      }
    }
  >
    Making Sense RN seed
  </Text>
  <Text
    style={
      Object {
        "color": "#9FA7BC",
        "fontSize": 17,
        "fontWeight": "400",
        "letterSpacing": 0.2,
        "marginBottom": 0,
      }
    }
  >
    Please login to your account
  </Text>
  <Text
    style={
      Object {
        "color": "#C21E5D",
        "fontSize": 13,
        "fontWeight": "400",
        "marginTop": 8,
      }
    }
  >
    Oops! Invalid email or password
  </Text>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginBottom": 110,
          "marginTop": 50,
        }
      }
    >
      <View
        style={
          Object {
            "marginBottom": 32,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9FA7BC",
              "fontSize": 13,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          EMAIL:
        </Text>
        <TextInput
          allowFontScaling={true}
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Enter your email..."
          placeholderTextColor="#D5DEF1"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              },
              Object {
                "borderColor": "#C21E5D",
                "borderWidth": 1,
              },
            ]
          }
          underlineColorAndroid="rgba(0, 0, 0, 0)"
          value="dgeslin@makingsense.com"
        />
      </View>
      <View
        style={
          Object {
            "marginBottom": 32,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9FA7BC",
              "fontSize": 13,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          PASSWORD:
        </Text>
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Enter your password..."
          placeholderTextColor="#D5DEF1"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              },
              Object {
                "borderColor": "#C21E5D",
                "borderWidth": 1,
              },
            ]
          }
          underlineColorAndroid="rgba(0, 0, 0, 0)"
          value="password"
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="login-button"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Login renders without crashing 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#1745D5",
        "fontSize": 23,
        "fontWeight": "600",
        "marginBottom": 8,
        "marginTop": 16,
        "textAlign": "left",
      }
    }
  >
    Making Sense RN seed
  </Text>
  <Text
    style={
      Object {
        "color": "#9FA7BC",
        "fontSize": 17,
        "fontWeight": "400",
        "letterSpacing": 0.2,
        "marginBottom": 0,
      }
    }
  >
    Please login to your account
  </Text>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginBottom": 110,
          "marginTop": 50,
        }
      }
    >
      <View
        style={
          Object {
            "marginBottom": 32,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9FA7BC",
              "fontSize": 13,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          EMAIL:
        </Text>
        <TextInput
          allowFontScaling={true}
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Enter your email..."
          placeholderTextColor="#D5DEF1"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              },
              false,
            ]
          }
          underlineColorAndroid="rgba(0, 0, 0, 0)"
          value="dgeslin@makingsense.com"
        />
      </View>
      <View
        style={
          Object {
            "marginBottom": 32,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9FA7BC",
              "fontSize": 13,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          PASSWORD:
        </Text>
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Enter your password..."
          placeholderTextColor="#D5DEF1"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              },
              false,
            ]
          }
          underlineColorAndroid="rgba(0, 0, 0, 0)"
          value="password"
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="login-button"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Login renders without crashing on Android 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFF",
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 24,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#1745D5",
        "fontSize": 23,
        "fontWeight": "600",
        "marginBottom": 8,
        "marginTop": 16,
        "textAlign": "left",
      }
    }
  >
    Making Sense RN seed
  </Text>
  <Text
    style={
      Object {
        "color": "#9FA7BC",
        "fontSize": 17,
        "fontWeight": "400",
        "letterSpacing": 0.2,
        "marginBottom": 0,
      }
    }
  >
    Please login to your account
  </Text>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "space-between",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginBottom": 110,
          "marginTop": 50,
        }
      }
    >
      <View
        style={
          Object {
            "marginBottom": 32,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9FA7BC",
              "fontSize": 13,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          EMAIL:
        </Text>
        <TextInput
          allowFontScaling={true}
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="Enter your email..."
          placeholderTextColor="#D5DEF1"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              },
              false,
            ]
          }
          underlineColorAndroid="rgba(0, 0, 0, 0)"
          value="dgeslin@makingsense.com"
        />
      </View>
      <View
        style={
          Object {
            "marginBottom": 32,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9FA7BC",
              "fontSize": 13,
              "fontWeight": "600",
              "marginBottom": 10,
            }
          }
        >
          PASSWORD:
        </Text>
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          placeholder="Enter your password..."
          placeholderTextColor="#D5DEF1"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              Object {
                "borderColor": "#D5DEF1",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#646565",
                "fontSize": 15,
                "fontWeight": "400",
                "padding": 16,
              },
              false,
            ]
          }
          underlineColorAndroid="rgba(0, 0, 0, 0)"
          value="password"
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderRadius": 3,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="login-button"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingBottom": 1,
              "paddingTop": 2,
              "textAlign": "center",
            }
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;
